javalib/src/main/scala/java/net/URLDecoder.scala

```diff
@@ -0,0 +1,78 @@
+package java.net
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.{Charset, MalformedInputException}
+
+object URLDecoder {
+
+  @Deprecated
+  def decode(s: String): String = decodeImpl(s, Charset.defaultCharset)
+
+  def decode(s: String, enc: String): String = {
+    /* An exception is thrown only if the
+     * character encoding needs to be consulted
+     */
+    lazy val charset = {
+      if (!Charset.isSupported(enc))
+        throw new UnsupportedEncodingException(enc)
+      else
+        Charset.forName(enc)
+    }
+
+    decodeImpl(s, charset)
+  }
+
+  private def decodeImpl(s: String, charset: => Charset): String = {
+
+    val len = s.length
+    lazy val charsetDecoder = charset.newDecoder()
+
+    lazy val byteBuffer = ByteBuffer.allocate(len / 3)
+    val charBuffer = CharBuffer.allocate(len)
+
+    lazy val illegalHex =
+      new IllegalArgumentException(
+          "URLDecoder: Illegal hex characters in escape (%) pattern")
+
+    var i = 0
+    while (i < len) {
+      s.charAt(i) match {
+        case '+' =>
+          charBuffer.append(' ')
+          i += 1
+
+        case '%' if i + 3 > len =>
+          throw illegalHex
+
+        case '%' =>
+          byteBuffer.clear()
+          charsetDecoder.reset()
```

Comment:
`charsetDecoder` is a `lazy val`. Accessing it every time has a cost. Consider doing `val decoder = charsetDecoder` above this line, and then use `decoder`.